spec/js-api/value/color.d.ts.md

```diff
@@ -39,83 +239,757 @@ The [private `internal` field] refers to a Sass color.
+
+    ```js
+    new SassColor({
+      red: options.red ?? color.channel('red'),
+      green: options.green ?? color.channel('green'),
+      blue: options.blue ?? color.channel('blue'),
+      alpha: options.alpha ?? color.channel('alpha'),
+      space: space
+    })
+    ```
+
+* If `space` equals `rgb` and `spaceSetExplicitly` is `true`, let `changedColor`
+  be the result of:
+
+   ```js
+  new SassColor({
+    red: changedValue('red'),
+    green: changedValue('green'),
+    blue: changedValue('blue'),
+    alpha: changedValue('alpha'),
+    space: space
+  })
+  ```
+
+* If `space` equals `lab` or `oklab`, let `changedColor` be the result of:
+
+  ```js
+  new SassColor({
+    lightness: changedValue('lightness'),
+    a: changedValue('a'),
+    b: changedValue('b'),
+    alpha: changedValue('alpha'),
+    space: space
+  })
+  ```
+
+* If `space` equals `lch` or `oklch`, let `changedColor` be the result of:
+
+  ```js
+  new SassColor({
+    lightness: changedValue('lightness'),
+    chroma: changedValue('chroma'),
+    hue: changedValue('hue'),
+    alpha: changedValue('alpha'),
+    space: space
+  })
+  ```
+
+* If `space` equals `a98-rgb`, `display-p3`, `prophoto-rgb`, `srgb`, or
+  `srgb-linear`, let `changedColor` be the result of:
+
+  ```js
+  new SassColor({
+    red: changedValue('red'),
+    green: changedValue('green'),
+    blue: changedValue('blue'),
+    alpha: changedValue('alpha'),
+    space: space
+  })
+  ```
+
+* If `space` equals `xyz`, `xyz-d50`, or `xyz-d65`, let `changedColor` be the
+  result of:
+
+  ```js
+  new SassColor({
+    y: changedValue('y'),
+    x: changedValue('x'),
+    z: changedValue('z'),
+    alpha: changedValue('alpha'),
+    space: space
+  })
+  ```
+
+* Return the result of [`changedColor.toSpace(initialSpace)`].
+
+[`changedColor.toSpace(initialSpace)`]: #tospace
+[`Changing a Component Value`]: #changing-a-component-value
+
+```ts
+change(
+  options: {
+    [key in ChannelNameHSL]?: number | null;
+  } & {
+    alpha?: number;
+    space?: ColorSpaceHSL;
+  }
+): SassColor;
+
+change(
+  options: {
+    [key in ChannelNameHWB]?: number | null;
+  } & {
+    alpha?: number;
+    space?: ColorSpaceHWB;
+  }
+): SassColor;
+
+change(
+  options: {
+    [key in ChannelNameLab]?: number | null;
+  } & {
+    alpha?: number | null;
+    space?: ColorSpaceLab;
+  }
+): SassColor;
+
+change(
+  options: {
+    [key in ChannelNameLCH]?: number | null;
+  } & {
+    alpha?: number | null;
+    space?: ColorSpaceLCH;
+  }
+): SassColor;
+
+change(
+  options: {
+    [key in ChannelNameRGB]?: number | null;
+  } & {
+    alpha?: number | null;
+    space?: ColorSpaceRGB;
+  }
+): SassColor;
+
+change(
+  options: {
+    [key in ChannelNameXYZ]?: number | null;
+  } & {
+    alpha?: number | null;
+    space?: ColorSpaceXYZ;
+  }
+): SassColor;
+```
+
 #### `red`
 
+> This is deprecated in favor of the new [`channel`] function. This deprecation
+> is called `color-4-api`.
```

Comment:
I wasn't sure what to do with deprecated getters -- is there an existing pattern for that?

Reply:
That's a good question. This is the first time we've deprecated part of the modern JS API. I think we have two options:

1. Treat it more like a Sass deprecation, log a warning when it's used, give that warning a name and allow it to be silenced like other deprecation warnings.

2. Treat it more like a normal JS deprecation, just mark it `@deprecated` and leave it up to users' tooling to notify them that they're using it.

The former provides a stronger guarantee that users will actually see it, but the latter is substantially easier to implement and is less obtrusive in some ways (for example, it can be silenced with standard eslint annotations). A quick search suggests that eslint doesn't actually surface warnings for deprecated member use by default, so it's probably safest to go with option 1.

Reply:
@nex3 Ah, I see. So what I have here is probably okay for the spec, but we'll need to figure out implementation.

My leaning would be option 2, since that's what JS developers would be used to (and matches how we're marking e.g. `renderSync` as deprecated). But if we want to consistently output to stderr even outside of a compile/logger context, that requires separate implementation.

Is this a new feature? That is, it sounds like maybe we need a new spec for what JS API deprecation does?